### ISSUE TITLE
Slight unsoundness when dealing with N-ZST pointers

### ISSUE THREAD

**Issue by giltho** [model-checking/kani#1341]
CBMC doesn't model pointer alignments, see #1168 . 
This leads to unsoundness in the kani proofs - although keep in mind that these are really corner cases.

Rust expects pointers to N-aligned ZSTs to have an alignment of N.
Creating a misaligned reference, or dereferencing a misaligned pointer is undefined behaviour. That is also the case with pointers to ZSTs, even though dereferencing a pointer to a ZST is no-op.

For example, this code is invalid Rust, and triggers undefined behaviour:
```rust
use std::ptr::NonNull;

#[derive(Clone, Copy)]
#[repr(align(32))]
struct Zst32;

#[kani::proof]
pub fn main() {
    unsafe {
        let u: NonNull<Zst32> = NonNull::new_unchecked(8 as *mut Zst32);
        let v = *(u.as_ptr());
    }
}
```
You can try [running it with MIRI in the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=use%20std%3A%3Aptr%3A%3ANonNull%3B%0A%0A%23%5Bderive(Clone%2C%20Copy)%5D%0A%23%5Brepr(align(32))%5D%0Astruct%20Zst32%3B%0A%0Apub%20fn%20main()%20%7B%0A%20%20%20%20unsafe%20%7B%0A%20%20%20%20%20%20%20%20let%20u%3A%20NonNull%3CZst32%3E%20%3D%20NonNull%3A%3Anew_unchecked(8%20as%20*mut%20Zst32)%3B%0A%20%20%20%20%20%20%20%20let%20v%20%3D%20*(u.as_ptr())%3B%0A%20%20%20%20%7D%0A%7D), it is properly caught: the alignment check should fail. However, Kani does not catch it.

---

Similarly, Rust guarantees that creating references to ZSTs will yield properly-aligned pointers. 
For example, the following code is a test that is guaranteed to always pass
```rust
#[repr(align(32))]
struct Zst32;

#[kani::proof]
pub fn main() {
    let u: Zst32 = Zst32;
    let v = &u;

    assert_eq!((v as *const Zst32 as usize) % 32, 0, "pointers to a 32-ZST should be 32 aligned")
}
```

But, this test should not pass:
```rust
#[repr(align(32))]
struct Zst32;

#[kani::proof]
pub fn main() {
    let u: Zst32 = Zst32;
    let v = &u;

    assert_eq!((v as *const Zst32 as usize) % 4096, 0, "pointers to a 32-ZST should be 4096 aligned - that is incorrect")
}
```
You can try [running it in the playground, with miri or not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=%23%5Brepr(align(32))%5D%0Astruct%20Zst32%3B%0A%0Apub%20fn%20main()%20%7B%0A%20%20%20%20let%20u%3A%20Zst32%20%3D%20Zst32%3B%0A%20%20%20%20let%20v%20%3D%20%26u%3B%0A%0A%20%20%20%20assert_eq!((v%20as%20*const%20Zst32%20as%20usize)%20%25%204096%2C%200%2C%20%22pointers%20to%20a%2032-ZST%20should%20be%204096%20aligned%20-%20that%20is%20incorrect%22)%0A%7D); it is a bit non-deterministic, but there is a very strong chance that it fails.
In kani, it passes. I've tried a few things, and it seems that pointers will behave as if they were aligned on any power of 2 in CBMC.

**Comment by celinval** [comment]
@tautschnig did your change to ZST fixed this?

**Comment by tautschnig** [comment]
Yes, this was indeed fixed in #3134, as the following test (extension of the original example) confirms:
````rust
#[repr(align(32))]
struct Zst32;

#[kani::proof]
pub fn main() {
    let u: Zst32 = Zst32;
    let v = &u;

    assert_eq!((v as *const Zst32 as usize) % 32, 0, "pointers to a 32-ZST should be 32 aligned");
    assert_eq!((v as *const Zst32 as usize) % 64, 0, "pointers to a 32-ZST should be 64 aligned - that is incorrect");
    assert_eq!((v as *const Zst32 as usize) % 4096, 0, "pointers to a 32-ZST should be 4096 aligned - that is incorrect")
}
````
The first assertion passes, the second and the third one fail as they should.
